### PR TITLE
8286030: Support sharing of /tmp across Linux containers

### PR DESCRIPTION
**Pull request by iklam**
There are some Kubernetes setups that share the same `/tmp` directory across multiple containers. Such a scenario is currently not supported by the JDK and crashes may happen.

### Goals

+ Allow JVM processes in different containers to share the `/tmp` directory
+ Serviceability tools such as `jps` and `jcmd` should be able to access all such processes.

### Non-goals

+ The `/tmp` directory can be shared only by updated JVMs with this patch. We do not support such sharing with older JVMs.
+ When using a shared `/tmp` directory, you must use the updated versions of `jps`, `jcmd`, etc, that contains this patch. Tools from older JDKs don't work.

### Compatibility

+ When the `/tmp` directory is NOT shared across containers, the behavior of the updated JVM is exactly the same as before. The updated JVM can co-exist with older JVMs, and can be accessed by tools like `jps`, `jcmd`, etc, from older JDKs.
+ The updated `jps`, `jcmd`, etc, tools in the updated JDK can work with JVMs from older JDKs.

### Design

AttachID -- this is the number used for `/tmp/hsperfdata_$USER/<n>` and `/tmp/.java_pid<n>`

When a JVM starts, it chooses its AttachID by this pseudo code:

```
    for (id = os::current_process_id(); ; id = random_id()) {
      if ((flock(/tmp/hsperfdata_$USER/$id) == SUCCESS) {
        break;
      }
    }
```

Note that `random_id()` always returns a negative integer, so we will not step on the `pid` of another valid process.

- See the LINUX version of `mmap_create_shared()` in this patch.

### Cleaning up unused perfdata files

See `unlink_sharedmem_file_if_stale()` in the patch. We use both `flock()` and `kill(pid, 0)` to check to liveness. This is a little complicated because we need to co-exist with older JVMs.

Note that if the `pid` is positive, it could be from an older JVM, which doesn't do `flock()`. Therefore, for such positive `pids`, we also use `kill(pid, 0)`, which is racy, but this is no worse than before.

### Discovering JVMs

+ See `LocalVmManager.activeVms()` -> `PlatformSupport.activeVms()`.
+ See `PlatformSupport.getAttachID()`.

The Linux version no longer scans `/tmp` directories. Instead, it scans for `hsperfdata` files in the `/proc/*/maps` files. This way, it can detect the AttachID of each JVM process. The AttachID may be randomized so it may be different than the JVM's host `pid` and `nspid`.

### Testing

I've done some quick manual testing and it seems to work when several containers are sharing the same `/tmp` directory. My command-line is this (`/tmp` inside the container is mapped from `/tmp/dockertmp` on the host file system):

```
docker run -it --tty=true --rm \
     -v /tmp/dockertmp:/tmp my-java-container \
     java -cp / -Xlog:perf+memops=debug -Xlog:attach=trace Wait
```

I'll try to write a jtreg test case. Since we can't run as root, I'll try to use podman and run the tests in rootless mode.

### About `-XX:+PerfDisableSharedMem`

With my patch, if you share `/tmp` across two containers and run `java -XX:+PerfDisableSharedMem ...` in both containers:

+ `jps` will not list either JVM process
+ `jcmd $HOSTPID` works for the first process that you connect to, but will fail with the second process. This is because the processes will try to use the same socket file, e.g., `/tmp/.java_pid1`

Both of the above issues are the same as before, so there's no regression. To avoid making this patch more complicated, I plan to fix this problem in a future bug by mapping an empty `/tmp/hsperfdata_$USER/$id` when `-XX:+PerfDisableSharedMem` is specified.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8286030](https://bugs.openjdk.org/browse/JDK-8286030): Support sharing of /tmp across Linux containers

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9226/head:pull/9226` \
`$ git checkout pull/9226`

Update a local copy of the PR: \
`$ git checkout pull/9226` \
`$ git pull https://git.openjdk.org/jdk pull/9226/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9226`

View PR using the GUI difftool: \
`$ git pr show -t 9226`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9226.diff">https://git.openjdk.org/jdk/pull/9226.diff</a>

</details>

### CODE REVIEW

**Comment by iklam** [comment]
TL/DR; This fix is too complex. I am opening a new PR for a simpler fix: https://github.com/openjdk/jdk/pull/9406

---

Since many people (or at least some of the customers that I heard about) are sharing /tmp in Kubernetes setups, I am surprised that we have not heard much of this kind of crash. I tried running a bunch of containers in parallel and even added random sleeps in the JVM. However, I couldn't reproduce the crash.

I ended up writing a manual test case to get the exact timing:

https://github.com/openjdk/jdk/compare/master...iklam:jdk:8286030-test-case-for-jvm-crash-when-containers-share-tmp-dir?expand=1

My understand of the problem is:

When two JVMs clash on the hsperfdata file name (the both try to use /tmp/hsperfdata_root/1):

- The first JVM creates the file and start using it.
- The second JVM, who also think its PID is 1, assumes that /tmp/hsperfdata_root/1 is a stale file (it assumes no other process can have the same PID as itself), so it deletes the file, and create a new one

Later, when you use jcmd, you can access only the second JVM, but in most of the time there's no crash.

However, on very rare conditions (with the exact timing of the above manual test), one of the JVMs might crash.

----

I want to change my plan:

(1) Since the crash problem is more catastrophic, it should be fixed first. This can be done quite easily with flock. This fix will not cure the jcmd problem, but it will not make the jcmd problem any worse than before.

(2) The jcmd problem is harder to fix. We could try something like this PR, or maybe try a different approach.

**Comment by iklam** [comment]
Some comments  about this PR from e-mail exchanges with @tstuefe:

> @tstuefe: Since negative pids are used for process groups, they can be valid and its possible to send signals to them. "-pid" means "group with process <pid> as leader". Make sure that you don't use kill for these pids. And that you comment this clearly, otherwise people may think you mean process groups. Matter of fact this also goes for old tools. If an old tool detects a "pid-123" file, would it send a signal to process group 123? Hm, this is tricky.
>
> Can you instead use a different scheme, e.g. an offset to push the pid beyond sizeof(int)?

Right. Maybe we can use something like hsperfdata_root/uid+1234

> @tstuefe: I am not sure how unlink_sharedmem_file_if_stale() is supposed to work. You are running in a VM inside a container, right? So, you only see the PIDs in your container. How can you use the PID for kill(3) then, since processes in neighboring containers are hidden from you?

I think we can do flock() first and then kill() second. If flock determines that the file is alive, then there's no need to call kill() anymore.

> @tstuefe:  Does flock(3) inside a container actually lock across containers and host?

According to my experiments on ubuntu, yes.

> @tstuefe:  Side note, skimming thru the code, I was unaware that the attacher creates a file inside the working directory of the attachee. I always thought this goes straight to tmp.

I was surprised, too. It seems like a hack to me ....
